File: 2015/2/elixir/main.exs
defmodule Main do
  {_, input} =
    File.read("../input.txt")

  input =
    input
    |> String.trim()
    |> String.split("\n")
    |> Enum.map(fn v ->
      v |> String.split("x") |> Enum.map(fn v -> String.to_integer(v) end)
    end)

  input
  |> Enum.reduce(0, fn v, acc ->
    [l, w, h] = v
    sides = [l * w, w * h, h * l]
    acc + Enum.sum(sides) * 2 + Enum.min(sides)
  end)
  |> IO.puts()

  input
  |> Enum.reduce(0, fn measure, acc ->
    [l, w, h] = measure
    half_perimeter = [l + w, w + h, h + l]

    acc + Enum.reduce(measure, fn a, b -> a * b end) +
      Enum.min(half_perimeter) * 2
  end)
  |> IO.puts()
end

File: 2015/2/rust/src/main.rs
use itertools::Itertools;

fn main() {
    let input =
        std::fs::read_to_string("../input.txt").expect("invalid input!!");
    let input = input
        .trim()
        .split("\n")
        .map(|v| v.split("x").map(|n| n.parse().expect("not a number")));

    let part1 = input.clone().fold(0, |acc, v| {
        let sides: Vec<i32> = v
            .combinations(2)
            .map(|pair| pair.into_iter().reduce(|a, b| a * b).unwrap())
            .collect();
        let sides = sides.into_iter();

        acc + sides.clone().fold(0, |acc, v| acc + v) * 2
            + sides.clone().reduce(|a, b| a.min(b)).unwrap()
    });
    println!("{part1}");

    let part2 = input.clone().fold(0, |acc, v| {
        let smallest_perimeter = v
            .clone()
            .map(|v| v * 2)
            .combinations(2)
            .map(|v| v.into_iter().fold(0, |a, b| a + b))
            .reduce(|a, b| a.min(b))
            .unwrap();

        acc + smallest_perimeter + v.fold(1, |a, b| a * b)
    });
    println!("{part2}");
}
